feat: hasMany for number field

AlessioGr opened the pull request:
## Description

This PR adds a `hasMany` option for the number field. Unlike the `array` field, this allows you to save a _clean_ number array in MongoDB.

The changes in react-select can also be useful for https://github.com/payloadcms/payload/pull/2347#issuecomment-1487432913

https://user-images.githubusercontent.com/70709113/232639795-f5ed05c1-5a98-45a0-b4e8-077a75f31b05.mp4


This can be used like this:

```ts
{
      name: 'multinumber',
      type: 'number',
      hasMany: true,
},
```

## Data structure comparison

### number field with hasMany:
![MongoDB Compass 2023-04-18 at 02 24 47@2x](https://user-images.githubusercontent.com/70709113/232638157-2433ab1a-7feb-4538-b8c2-3a043b4b3022.jpg)

### array field with number fields:
![MongoDB Compass 2023-04-18 at 02 25 18@2x](https://user-images.githubusercontent.com/70709113/232638185-af0d6d91-2056-412b-bead-02c273810034.jpg)


- [X] I have read and understand the CONTRIBUTING.md document in this repository

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Existing test suite passes locally with my changes
- [ ] I have made corresponding changes to the documentation

DanRibbens commented:
I like what I see here!

DanRibbens reviewed (changes requested):
I'm good with these changes. If you can fix the tab issue I mentioned and remove the console logs then I think this is done.

AlessioGr commented:
Thanks, Dan, for reviewing the PR! I'm a bit pressed for time at the moment, but I'll make sure to address the requested changes as soon as I get a chance!

Quick note to self: I may need to tweak the validation. It seems to run into issues when I add a value, save it, and then clear all the values.

AlessioGr commented:
@DanRibbens this is ready for review again!

DanRibbens commented on src/fields/validations.ts:
```diff
@@ -24,25 +24,34 @@ import canUseDOM from '../utilities/canUseDOM';
+
+  // eslint-disable-next-line no-restricted-syntax
+  for (const val of toValidate) {
+    const v = parseFloat(val as unknown as string);
```
Try to avoid single character variable names if you can help it. Even in a simple function like this, it isn't as readable.

DanRibbens commented on src/admin/components/elements/ReactSelect/index.tsx:
```diff
@@ -41,8 +52,55 @@ const SelectAdapter: React.FC<ReactSelectAdapterProps> = (props) => {
+    switch (event.key) {
+      case 'Enter':
+      case 'Tab':
+        onChange([...value as Option[], createOption(inputValue)]);
```
When I type something invalid and press tab, it will enter "NaN". Ideally, you wouldn't even be able to enter any string and the input would behave like a `type="number"`. Is that possible with react-select the way we are using it?

It wasn't clear in their docs if this is possible: https://react-select.com/components#:~:text=Yellow-,Input,-Input%20to%20render

DanRibbens reviewed (commented):
Good work! I think we're very close. I gave a few nice-to-have comments to see if you have some thoughts. These aren't deal-breakers, If they are easy improvements it is worthwhile, otherwise we can revisit later.

JarrodMFlesch commented:
## GraphQL generation updates
- [buildMutationInputType](https://github.com/payloadcms/payload/blob/master/src/graphql/schema/buildMutationInputType.ts#L42)
- [withOperators](https://github.com/payloadcms/payload/blob/master/src/graphql/schema/withOperators.ts#L29)

## TS generation updates
- [entityToJSONSchema](https://github.com/payloadcms/payload/blob/master/src/graphql/schema/withOperators.ts#L29)

@jmikrut can you think of any others?

AlessioGr commented on src/admin/components/forms/field-types/Relationship/types.ts:
```diff
@@ -10,7 +10,7 @@ export type Props = Omit<RelationshipField, 'type'> & {
 export type Option = {
   label: string
-  value: string | number
+  value: string
```
Not having it as a string would break sortable react-selects. I added number back and fixed it by stringifying the value in react-select using `''+value`, though. Added it back, should be fine now!

AlessioGr commented:
@DanRibbens I've fixed all open requested changes!

@JarrodMFlesch I've added the updated types to [entityToJSONSchema](https://github.com/payloadcms/payload/blob/master/src/utilities/entityToJSONSchema.ts#L68) as well as [buildMutationInputType](https://github.com/payloadcms/payload/blob/master/src/graphql/schema/buildMutationInputType.ts#L42)

I'm not sure about [withOperators](https://github.com/payloadcms/payload/blob/master/src/graphql/schema/withOperators.ts#L29). What does `defaults` exactly do? It looks like that's the type if you added no special config to the field (e.g. `hasMany`) - in that case, it would be correct leaving it like that.

JarrodMFlesch commented:
> @JarrodMFlesch I've added the updated types to [entityToJSONSchema](https://github.com/payloadcms/payload/blob/master/src/utilities/entityToJSONSchema.ts#L68) as well as [buildMutationInputType](https://github.com/payloadcms/payload/blob/master/src/graphql/schema/buildMutationInputType.ts#L42)
> 
> I'm not sure about [withOperators](https://github.com/payloadcms/payload/blob/master/src/graphql/schema/withOperators.ts#L29). What does `defaults` exactly do? It looks like that's the type if you added no special config to the field (e.g. `hasMany`) - in that case, it would be correct leaving it like that.

The end goal would be to have `hasMany` return the same shape as the point field [seen here](https://github.com/payloadcms/payload/blob/master/src/graphql/schema/withOperators.ts#L85).

You can turn it into a function that takes the field as an argument, just like the radio field, then check if it is hasMany and adjust the returned type based on that.

AlessioGr commented:
Hm shouldn't we do the same for `relationship`? In defaults, it's currently just `GraphQLString` even though it could be an array of strings.

Also, found another place where I've updated the type accordingly:
https://github.com/AlessioGr/payload/blob/number-hasmany-v2/src/graphql/schema/buildObjectType.ts#L87

JarrodMFlesch commented on src/graphql/schema/withOperators.ts:
```diff
@@ -27,7 +27,10 @@ type DefaultsType = {
   number: {
-    type: GraphQLFloat,
+    type: (field: NumberField): GraphQLType => {
+      const type = field?.name === 'id' ? GraphQLInt : GraphQLFloat;
```
Good catch 👍

JarrodMFlesch commented:
> Hm shouldn't we do the same for `relationship`? In defaults, it's currently just `GraphQLString` even though it could be an array of strings.
> 
> Also, found another place where I've updated the type accordingly: https://github.com/AlessioGr/payload/blob/number-hasmany-v2/src/graphql/schema/buildObjectType.ts#L87

Yeah, I think we should update that 👍

AlessioGr commented:
> > Hm shouldn't we do the same for `relationship`? In defaults, it's currently just `GraphQLString` even though it could be an array of strings.
> > Also, found another place where I've updated the type accordingly: https://github.com/AlessioGr/payload/blob/number-hasmany-v2/src/graphql/schema/buildObjectType.ts#L87
> 
> Yeah, I think we should update that 👍

Alright, updated it! Seems to be correct in other places

JarrodMFlesch commented:
The only thing I want to bring up before merging this is the usage of `min` & `max` when `hasMany: true`. It is interesting because we use `min` & `max` on relationship fields and it has a different meaning - the length of the array. I am wondering if this will become confusing into the future? Would someone expect min and max to affect the number of values to be stored in the array?

You can see how relationship field validates currently [here](https://github.com/payloadcms/payload/pull/2517/files#diff-e8686d63ee7a57e9238921baf4b95b80322b8df7d080f96d305fad65940678f8L292).

AlessioGr commented:
What about

```ts
hasMany: boolean | {
  min: number,
  max: number
}
```
?

Would allow you to customize the minimum and maximum amount of numbers. But, it would still be inconsistent with what relationship does

JarrodMFlesch commented:
> What about
> 
> ```ts
> hasMany: boolean | {
>   min: number,
>   max: number
> }
> ```
> 
> ?
> 
> Would allow you to customize the minimum and maximum amount of numbers. But, it would still be inconsistent with what relationship does

Haha! I was just typing that out, I think that is the move!

DanRibbens commented:
I thought about the `min` and `max` validation when reviewing also. IMO an added feature for validation options would be to add `maxRows` and `minRows` which is what we call that in other fields.

DanRibbens commented on test/fields/int.spec.ts:
```diff
@@ -194,6 +194,50 @@ describe('Fields', () => {
       })).rejects.toThrow('The following field is invalid: decimalMax');
     });
+
+    it('should validate an array of numbers', async () => {
```
Move validation tests to `src/field/validations.spec.ts` since they don't need to be integration.

DanRibbens reviewed (changes requested):
I have an issue with the frontend validation.
While the enters, we are running the validation against one value:
![image](https://github.com/payloadcms/payload/assets/6434612/6e0e0b56-f91d-4557-8528-215fa8a526f0)

But when you save it runs with the value as an array:
![image](https://github.com/payloadcms/payload/assets/6434612/76e7d00a-a219-40e9-9a79-767fc734601c)

We need to call the validation function with a consistent shape. I'm thinking that means it has to be the full array and maybe it isn't called until the user clicks "Create". What do you think?

AlessioGr commented:
> I have an issue with the frontend validation. While the enters, we are running the validation against one value: ![image](https://user-images.githubusercontent.com/6434612/246468792-6e0e0b56-f91d-4557-8528-215fa8a526f0.png)
> 
> But when you save it runs with the value as an array: ![image](https://user-images.githubusercontent.com/6434612/246469390-76e7d00a-a219-40e9-9a79-767fc734601c.png)
> 
> We need to call the validation function with a consistent shape. I'm thinking that means it has to be the full array and maybe it isn't called until the user clicks "Create". What do you think?

@DanRibbens Yep you're completely right! I've just updated that with the last commit! Now, as long as hasMany is true, it's always an array of numbers.

I'd still keep it that way where it validates every time you start typing, as it allows you to catch issues way earlier - by not letting you insert incorrect values in the first place

JarrodMFlesch commented on src/admin/components/forms/field-types/Number/index.tsx:
```diff
@@ -145,7 +145,7 @@ const NumberField: React.FC<Props> = (props) => {
           filterOption={(option, rawInput) => {
             // eslint-disable-next-line no-restricted-globals
             const notANumber = rawInput.trim() === '' || isNaN(Number(rawInput));
```
Maybe we should use this [isNumber](https://github.com/payloadcms/payload/blob/master/src/utilities/isNumber.ts) that we are using elsewhere?

AlessioGr replied:
@JarrodMFlesch Done! I've adjusted the isNumber function as well, as it didn't handle the edge-case where JavaScript is dumb and converts "   " into 0

DanRibbens commented:
> I'd still keep it that way where it validates every time you start typing, as it allows you to catch issues way earlier - by not letting you insert incorrect values in the first place

The only thing is that you don't get feedback and it might not be obvious to the user why a value is not creating without showing the validation response.

AlessioGr commented:
> > I'd still keep it that way where it validates every time you start typing, as it allows you to catch issues way earlier - by not letting you insert incorrect values in the first place
> 
> The only thing is that you don't get feedback and it might not be obvious to the user why a value is not creating without showing the validation response.

hm yeah you're right about that. Should I remove that part completely? I'd still keep the isNumber check in the filterOptions, since that part should be obvious

AlessioGr commented:
Should be good now, I've replaced the filterOptions with a simple isNumber check

DanRibbens reviewed (changes requested):
Looks great! Pulled it down and I can't find anything else wrong. I just found one docs typo for you to change and then this should get merged and released.

AlessioGr commented on docs/fields/number.mdx:
```diff
@@ -18,6 +18,9 @@ keywords: number, fields, config, configuration, documentation, Content Manageme
 | **`max`**            | Maximum value accepted. Used in the default `validation` function. |
+| **`hasMany`**        | Makes this field an ordered array of numbers instead of just a single number. |
+| **`minRows`**        | Minimum number of numbers in the numbers array, if `hasMany` is set to true. |
+| **`minRows`**        | Maximum number of numbers in the numbers array, if `hasMany` is set to true. |
```
Whoops. Fixed it!
